src/lib/storage/disk_analyzer.rb

```diff
@@ -0,0 +1,368 @@
+          installation_disks << disk if installation_disk?(disk)
+          disk_count += 1
+          if disk_count > @disk_check_limit
+            log.info("Disk check limit reached after #{disk}")
+            break
+          end
+        end
+        installation_disks
+      end
+
+      # Find disks that are suitable for installing Linux.
+      # Put any USB disks to the end of that array.
+      #
+      # @return [Array<::Storage::Disk>] candidate disks
+      #
+      def find_candidate_disks
+        disks = StorageManager.instance.probed.all_disks.to_a
+        # FIXME: to_a should not be necessary
+
+        usb_disks, non_usb_disks = disks.partition { |disk| disk.transport == ::Storage::USB }
+        log.info("USB Disks:     #{dev_names(usb_disks)}")
+        log.info("Non-USB Disks: #{dev_names(non_usb_disks)}")
+
+        # We don't want to install on our installation disk if there is any other way.
+        candidate_disks = remove_installation_disks(non_usb_disks + usb_disks)
+        if candidate_disks.empty?
+          log.info("No candidate disks left after eliminating installation disks")
+          log.info("Trying with non-USB installation disks")
+          candidate_disks = @installation_disks.select { |disk| disk.transport != ::Storage::USB }
+        end
+        if candidate_disks.empty?
+          log.info("Still no candidate disks left")
+          log.info("Trying with installation disks out of sheer desperation")
+          candidate_disks = @installation_disks
+        end
+        candidate_disks
+      end
+
+      # Remove any installation disks from 'disks' and return a disks array
+      # containing the disks that are not installation media.
+      #
+      # @param disks [Array<::Storage::Disk>]
+      # @return [Array<::Storage::Disk>] non-installation disks
+      #
+      def remove_installation_disks(disks)
+        # We can't simply use
+        #   disks -= @installation_disks
+        # because the list elements (from libstorage) don't provide a .hash method.
+        # Comparing device names ("/dev/sda"...) instead.
+
+        inst_names = dev_names(@installation_disks)
+        disks.delete_if { |disk| inst_names.include?(disk.name) }
+      end
+
+      # Find any Linux partitions on any of the candidate disks.
+      # This may be a normal Linux partition (type 0x83), a Linux swap
+      # partition (type 0x82), an LVM partition, or a RAID partition.
+      #
+      # @return [Array<Partition>] Linux partitions
+      #
+      def find_linux_partitions
+        linux_partitions = []
+        @candidate_disks.each do |disk|
+          begin
+            disk.partition_table.partitions.each do |partition|
+              if LINUX_PARTITION_IDS.include?(partition.id)
+                log.info("Found Linux partition #{partition} (ID 0x#{partition.id.to_s(16)})")
+                linux_partitions << partition
+              end
+            end
+          rescue RuntimeError => ex
```

Comment:
Consider it if this is your salary and function count number of your working days. If some day contain problem ( like you fill something wrong ), then you do not get salary for that day without any error message. At least me it cause very unhappy and I really prefer to know, that there is problem with counting my working days rather then silently decrease my salary.